File: go/fetchers/vps.go
package fetchers

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/lotusmarket/lotusmarket-go/types"
)

const vpsBaseURL = "https://bgapidatafeed.vps.com.vn/getliststockdata"

func VPS(ctx context.Context, ticker string) (*types.StockData, error) {
	stocks, err := VPSMultiple(ctx, []string{ticker})
	if err != nil {
		return nil, err
	}
	if len(stocks) == 0 {
		return nil, fmt.Errorf("no data for %s", ticker)
	}
	return &stocks[0], nil
}

func VPSMultiple(ctx context.Context, tickers []string) ([]types.StockData, error) {
	url := vpsBaseURL + "/" + strings.Join(tickers, ",")
	client := &http.Client{Timeout: 15 * time.Second}
	var lastErr error
	for attempt := 0; attempt < 3; attempt++ {
		if attempt > 0 {
			time.Sleep(time.Duration(attempt) * time.Second)
		}
		req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
		if err != nil {
			return nil, fmt.Errorf("create request: %w", err)
		}
		resp, err := client.Do(req)
		if err != nil {
			lastErr = err
			continue
		}
		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			lastErr = err
			continue
		}
		return parseVPSResponse(body)
	}
	return nil, fmt.Errorf("VPS API failed after 3 attempts: %w", lastErr)
}

func parseVPSResponse(data []byte) ([]types.StockData, error) {
	var raw []map[string]interface{}
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}
	var stocks []types.StockData
	for _, item := range raw {
		ticker := getString(item, "sym")
		if ticker == "" {
			continue
		}
		refPrice := getFloat(item, "r")
		lastPrice := getFloat(item, "lastPrice")
		openPrice := getFloat(item, "openPrice")
		if lastPrice == 0 && refPrice > 0 {
			lastPrice = refPrice
		}
		var changeValue, changePc float64
		if refPrice > 0 {
			changeValue = lastPrice - refPrice
			changePc = (changeValue / refPrice) * 100
		}
		stocks = append(stocks, types.StockData{
			Ticker:         ticker,
			Date:           time.Now().Format("2006-01-02"),
			Open:           openPrice,
			High:           getFloat(item, "highPrice"),
			Low:            getFloat(item, "lowPrice"),
			Close:          lastPrice,
			Volume:         getInt(item, "lot"),
			ChangePercent:  changePc,
			ChangeValue:    changeValue,
			RefPrice:       refPrice,
			Ceiling:        getFloat(item, "c"),
			Floor:          getFloat(item, "f"),
			ForeignBuyVol:  getInt(item, "fBVol"),
			ForeignSellVol: getInt(item, "fSVolume"),
			ForeignNetVol:  getInt(item, "fBVol") - getInt(item, "fSVolume"),
			Bid:            parseFloatStr(getString(item, "g1")),
			Ask:            parseFloatStr(getString(item, "g2")),
			BidVol:         parsePipeVol(getString(item, "g1")),
			AskVol:         parsePipeVol(getString(item, "g2")),
		})
	}
	for i := range stocks {
		stocks[i].ToVND()
	}
	return stocks, nil
}

File: python/lotusmarket/fetchers/vps.py
"""VPS DataFeed API — real-time quotes (no API key needed)."""

import json
from datetime import date
from typing import List
from lotusmarket.types import StockData

VPS_BASE_URL = "https://bgapidatafeed.vps.com.vn/getliststockdata"


def _parse_pipe_vol(s: str) -> int:
    parts = s.split("|")
    if len(parts) >= 2:
        try:
            return int(parts[1])
        except ValueError:
            return 0
    return 0


def parse_vps_response(data: bytes) -> List[StockData]:
    """Parse raw VPS JSON response into StockData list (prices in VND)."""
    raw = json.loads(data)
    stocks = []
    for item in raw:
        ticker = str(item.get("sym", ""))
        if not ticker:
            continue
        ref_price = float(item.get("r", 0))
        last_price = float(item.get("lastPrice", 0))
        open_price = float(item.get("openPrice", 0))
        if last_price == 0 and ref_price > 0:
            last_price = ref_price
        change_value = 0.0
        change_pct = 0.0
        if ref_price > 0:
            change_value = last_price - ref_price
            change_pct = (change_value / ref_price) * 100
        f_buy = int(float(item.get("fBVol", 0)))
        f_sell = int(float(item.get("fSVolume", 0)))
        g1 = str(item.get("g1", ""))
        g2 = str(item.get("g2", ""))
        s = StockData(
            ticker=ticker,
            date=date.today().isoformat(),
            open=open_price,
            high=float(item.get("highPrice", 0)),
            low=float(item.get("lowPrice", 0)),
            close=last_price,
            volume=int(float(item.get("lot", 0))),
            change_percent=change_pct,
            change_value=change_value,
            ref_price=ref_price,
            ceiling=float(item.get("c", 0)),
            floor=float(item.get("f", 0)),
            foreign_buy_vol=f_buy,
            foreign_sell_vol=f_sell,
            foreign_net_vol=f_buy - f_sell,
            bid=float(g1.split("|")[0]) if "|" in g1 else 0.0,
            ask=float(g2.split("|")[0]) if "|" in g2 else 0.0,
            bid_vol=_parse_pipe_vol(g1),
            ask_vol=_parse_pipe_vol(g2),
        )
        s.to_vnd()
        stocks.append(s)
    return stocks


def vps(ticker: str) -> StockData:
    """Fetch real-time quote for a single ticker."""
    stocks = vps_multiple([ticker])
    if not stocks:
        raise ValueError(f"No data for {ticker}")
    return stocks[0]


def vps_multiple(tickers: List[str]) -> List[StockData]:
    """Fetch real-time quotes for multiple tickers. Retries 3 times."""
    import httpx

    url = f"{VPS_BASE_URL}/{','.join(tickers)}"
    last_err = None
    for attempt in range(3):
        try:
            resp = httpx.get(url, timeout=15)
            resp.raise_for_status()
            return parse_vps_response(resp.content)
        except Exception as e:
            last_err = e
    raise ConnectionError(f"VPS API failed after 3 attempts: {last_err}")
